### PR TITLE
Add MatrixFreeOperators::BlockBase

### PR DESCRIPTION
**Pull request by masterleinad**
This PR adds the same functionality as MatrixFreeOperators::Base but for BlockVectorTypes.

The implementation is basically identical to what we do in Base but for each component of a BlockVector.
Therefore, `edge_constrained_indices` and `edge_constrained_values` are now `std::vectors` with entries for each block.

The Multigrid part is not yet tested.

### CODE REVIEW

**Comment by kronbichler** [comment]
Before looking into the specifics of the code I have a general comment: Given that the declaration of `MatrixFreeOperators::BlockBase` shares 90%, maybe more, of code with `MatrixFreeOperators::Base` in the new templated format, I wonder whether we should just use a single class. The only thing that really differs is the fact that we have a `std::vector` of `MGConstrainedDoFs` objects coming in to `initialize` and some internal storage of `std::vector<std::vector<...> >`. While the former could be solved by a different constructor, the latter could be realized by simply extending the non-block case with an outer std::vector of size 1. Wouldn't everything become much more compact then? One might need the template magic `internal::bool2type<IsBlockVector<VectorType>::value` to distinguish them, but I don't like copy-pasting interfaces 300 lines of interfaces and 500+ lines of code and rather prefer 10 or 20 lines of complicated template deduction stuff inside the library.

**Comment by davydden** [comment]
fully agree with @kronbichler , we should find a way to reuse the `MatrixFreeOperators::Base`, especially given the fact that it's now templated with the vector type.

**Comment by masterleinad** [comment]
That's fine with me. So I am going to make sure that `MatrixFree::Base` also works with the interface of `MatrixFree::BlockBase`

**Comment by davydden** [comment]
btw, this block operator is intended to be an operation on a single block of the block system, right?

**Comment by tamiko** [comment]
We use the terminology "block operator" in the library to describe an operator acting on full blocked vectors (with possible access to the individual operator acting on a single block). I guess we do the same here.

**Comment by masterleinad** [comment]
Yes, this is meant to be an operator that acts on all the blocks. 
An example would be a Stokes operator based on this class which I will add later as a test.

**Comment by masterleinad** [comment]
This is as close as I seem to get dealing with both Vectors and BlockVectors in the same class.

File: include/deal.II/matrix_free/operators.h [hunk 1]
```diff
@@ -35,6 +35,345 @@ DEAL_II_NAMESPACE_OPEN
...
+    (std::vector<std::vector<std::pair<typename VectorType::value_type,
+     typename VectorType::value_type> >> &edge_constrained_values,
+     const std::vector<std::vector<unsigned int> > &edge_constrained_indices,
+     VectorType &src,
+     const VectorType &dst)
+    {
+      typedef typename VectorType::value_type Number;
+      AssertDimension(edge_constrained_values.size(), 1);
+      AssertDimension(edge_constrained_indices.size(), 1);
+      for (unsigned int j = 0; j < edge_constrained_indices[0].size(); ++j)
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: do you need to have this one? If non-block vector returns `n_blocks()==1`, then the function above will be equivalent, right?

Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: same below.

Comment (masterleinad) [inline include/deal.II/matrix_free/operators.h]: In the function above, I access a block of the vector which I can't for non-block vectors. Therefore, I don't immediately see how to avoid splitting this function.

File: include/deal.II/matrix_free/operators.h [hunk 2]
```diff
@@ -828,29 +1194,28 @@ namespace MatrixFreeOperators
   Base<dim,VectorType>::vmult (VectorType       &dst,
                                const VectorType &src) const
   {
-    typedef typename Base<dim,VectorType>::value_type Number;
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: why this change? We don't use vectors this way now, but I think your code won't work with `std::complex<double>`.

**Comment by masterleinad** [comment]
The naming for all the `external` functions that don't have a direct counterpart in `MatrixFreeBase` can clearly be improved.

File: include/deal.II/matrix_free/operators.h (hunk 1 above)
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: right, i missed that `src.block()` thing... Since the difference between those functions are, in many cases, in 
`src.block(i)` vs `src`, why don't we put these accessors as a templates and then probably all the functions will be equivalent as long as you can get that non-block vector has a single block?

Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: what i mean is something like
```
     whatever my_block = internal_get_block(vec, i);
     my_block.local_element() = whatever
```
This is somewhat an alternative to making non-block vectors provide `block()` function which returns the vector itself and only accepts `0` as a block index.

**Comment by davydden** [comment]
@kronbichler @masterleinad 

> making non-block vectors provide block() function which returns the vector itself and only accepts 0 as a block index.

maybe we should just do that?

File: include/deal.II/matrix_free/operators.h (hunk 1 above)
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: or use this straight away:
```
internal_get_block(vec, i).local_element() = whatever
```
then it's only a matter of making `internal_get_block` return the vector for non-block vectors and required block for block vectors.

File: include/deal.II/matrix_free/operators.h [hunk 3]
```diff
@@ -35,6 +35,345 @@ DEAL_II_NAMESPACE_OPEN
...
+  namespace
+  {
+    template <int dim, typename VectorType>
+    typename std_cxx11::enable_if<IsBlockVector<VectorType>::value, void>::type
+    external_initialize_dof_vector(const MatrixFree<dim, typename VectorType::value_type> &mf,
+                                   VectorType &vec)
+    {
+      for (unsigned int i = 0; i < vec.n_blocks(); ++i)
+        {
+          if (!vec.block(i).partitioners_are_compatible(*mf.get_dof_info(i).vector_partitioner))
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: looking again at all these auxiliary functions, i think we can make them go by
```
for (unsigned int i = 0; i < internal_n_blocks(vec); ++i)
   {
        internal_get_block(vec,i).whatever...
   }
```
The difference appears to be only in accessing the vector directly or it's block.

Comment (masterleinad) [inline include/deal.II/matrix_free/operators.h]: I'll give it a shot :-)

Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: thanks! I hope it will work out and result in a cleaner code.

File: include/deal.II/matrix_free/operators.h (hunk 1 above)
Comment (kronbichler) [inline include/deal.II/matrix_free/operators.h]: Just as a remark, have a look how this is solved in `fe_evaluation.h`: https://github.com/dealii/dealii/blob/master/include/deal.II/matrix_free/fe_evaluation.h#L2868
Maybe you can even re-use the same data structure.

**Comment by masterleinad** [comment]
@kronbichler I followed a similar approach to what you have done in `fe_evaluation.h`.
@davydden I guess this is much cleaner and we are down to about 300 changed lines.

All `matrix_free` tests are passing.

File: include/deal.II/matrix_free/operators.h [hunk 4]
```diff
@@ -860,26 +975,31 @@ namespace MatrixFreeOperators
...
-    const_cast<VectorType &>(src).
-    reinit(data->get_dof_info(0).vector_partitioner);
-    VectorView<Number> view_src_out(src.local_size(), src.begin());
-    static_cast<Vector<Number>&>(view_src_out) = copy_vec;
+    for (unsigned int i=0; i<n_blocks(src); ++i)
+      {
+        // If both vectors use the same partitioner -> done
+        if (subblock(src,i).get_partitioner().get() ==
+            data->get_dof_info(i).vector_partitioner.get())
+          return;
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: probably `continue` here, otherwise you may `return` on the first block?

**Comment by masterleinad** [comment]
This should be it. I fixed that one remainig issue and added one test using an adaptive mesh and only hanging node constraints.

**Comment by kronbichler** [comment]
I have a problem with the new `Base` operator: Assume I have a `MatrixFree` that involves more than one component. I typically use this when I use `MatrixFree` in problems with inhomogeneous boundary conditions: Some evaluations go with non-zero boundary conditions interpolated while others use zero boundary conditions, the latter is in particular used while solving the linear system.

I would suggest we add two optional `ComponentMask` objects to the three initialize functions, one for the rows and one for the columns that let the user select a certain number of components, including a rectangular version. At this point, the selection made in the operator and what is actually used by the user needs to be done at two places but that is not possible without a major redesign of the loop that abstracts away the vector access anyway.

**Comment by kronbichler** [comment]
@davydden @masterleinad if you agree on this interface, I will create a PR.

**Comment by davydden** [comment]
@kronbichler sounds good to me. I use rather simple scalar valued operators (like Laplace), but a student of ours use vector-valued elasticity, so as long as the default behaviour is unchanged, I agree to any modifications needed to extend the `Base` operator to cover large array of usage cases.
